File: Graph_Traversal/BOJ_2606(바이러스_dfs).cpp
#include <iostream>
#include <vector>
#include <algorithm>
#define MAX 101
using namespace std;

vector<int> graph[MAX];
bool visited[MAX];
int n, m, k, a, b;
int cnt = 0;

int dfs(int start) {
	visited[start] = true;
	for (int i = 0; i < graph[start].size(); i++) {
		int next = graph[start][i];
		if (!visited[next]) {
			dfs(next);
			cnt += 1;
		}
	}
	return cnt;
}

int main(void) {
	cin >> n;
	cin >> m;
	for (int i = 0; i < m; i++) {
		cin >> a >> b;
		graph[a].push_back(b);
		graph[b].push_back(a);
	}
	for (int i = 1; i <= n; i++) {
		sort(graph[i].begin(), graph[i].end());
	}
	cout << dfs(1);
}

File: Graph_Traversal/BOJ_2606(바이러스_dfs).py
#DFS풀이
n=int(input())
m=int(input())
graph=[[]*i for i in range(n+1)]
visited=[False]*(n+1)
answer=[]
for _ in range(m): #그래프 생성
    a,b=map(int,input().split())
    graph[a].append(b)
    graph[b].append(a)
    graph[a].sort()
    graph[b].sort()
def dfs(graph,start,visited,answer):
    visited[start]=True
    for i in graph[start]:
        if visited[i]==False:
            answer.append(i)
            dfs(graph,i,visited,answer)
    return len(answer)
print(dfs(graph,1,visited,answer))
